File: capi/controller/internal/resources/access.go
package resources

import (
	rbacv1 "k8s.io/api/rbac/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
)

func BootstrapRBAC() []runtime.Object {
	subjects := []rbacv1.Subject{
		{APIGroup: rbacv1.GroupName, Kind: "Group", Name: "system:bootstrappers:kubeadm:default-node-token"},
		{APIGroup: rbacv1.GroupName, Kind: "Group", Name: "system:nodes"},
	}
	result := make([]runtime.Object, 0, 4)
	for _, item := range []struct {
		name     string
		resource string
	}{
		{name: "kubeadm:nodes-kubeadm-config", resource: "kubeadm-config"},
		{name: "kubeadm:kubelet-config", resource: "kubelet-config"},
	} {
		result = append(result,
			&rbacv1.Role{
				TypeMeta:   metav1.TypeMeta{APIVersion: rbacv1.SchemeGroupVersion.String(), Kind: "Role"},
				ObjectMeta: metav1.ObjectMeta{Name: item.name, Namespace: "kube-system"},
				Rules: []rbacv1.PolicyRule{{
					APIGroups:     []string{""},
					Resources:     []string{"configmaps"},
					ResourceNames: []string{item.resource},
					Verbs:         []string{"get"},
				}},
			},
			&rbacv1.RoleBinding{
				TypeMeta:   metav1.TypeMeta{APIVersion: rbacv1.SchemeGroupVersion.String(), Kind: "RoleBinding"},
				ObjectMeta: metav1.ObjectMeta{Name: item.name, Namespace: "kube-system"},
				RoleRef: rbacv1.RoleRef{
					APIGroup: rbacv1.GroupName,
					Kind:     "Role",
					Name:     item.name,
				},
				Subjects: subjects,
			},
		)
	}
	return result
}

File: capi/controller/src/resources/access.rs
use k8s_openapi::{
    api::rbac::v1::{PolicyRule, Role, RoleBinding, RoleRef, Subject},
    apimachinery::pkg::apis::meta::v1::ObjectMeta,
};

#[derive(Clone, Debug)]
pub struct BootstrapRole {
    pub role: Role,
    pub binding: RoleBinding,
}

pub fn bootstrap_rbac() -> Vec<BootstrapRole> {
    let subjects = [
        "system:bootstrappers:kubeadm:default-node-token",
        "system:nodes",
    ]
    .map(|name| Subject {
        api_group: Some("rbac.authorization.k8s.io".into()),
        kind: "Group".into(),
        name: name.into(),
        ..Default::default()
    })
    .to_vec();
    [
        ("kubeadm:nodes-kubeadm-config", "kubeadm-config"),
        ("kubeadm:kubelet-config", "kubelet-config"),
    ]
    .into_iter()
    .map(|(name, resource)| {
        let metadata = ObjectMeta {
            name: Some(name.into()),
            namespace: Some("kube-system".into()),
            ..Default::default()
        };
        BootstrapRole {
            role: Role {
                metadata: metadata.clone(),
                rules: Some(vec![PolicyRule {
                    api_groups: Some(vec!["".into()]),
                    resources: Some(vec!["configmaps".into()]),
                    resource_names: Some(vec![resource.into()]),
                    verbs: vec!["get".into()],
                    ..Default::default()
                }]),
            },
            binding: RoleBinding {
                metadata,
                role_ref: RoleRef {
                    api_group: Some("rbac.authorization.k8s.io".into()),
                    kind: "Role".into(),
                    name: name.into(),
                },
                subjects: Some(subjects.clone()),
            },
        }
    })
    .collect()
}

pub fn bootstrap_subjects_match(actual: &[Subject], expected: &[Subject]) -> bool {
    let key = |subject: &Subject| {
        (
            subject.api_group.clone().unwrap_or_default(),
            subject.kind.clone(),
            subject.name.clone(),
            subject.namespace.clone().unwrap_or_default(),
        )
    };
    let mut actual: Vec<_> = actual.iter().map(key).collect();
    let mut expected: Vec<_> = expected.iter().map(key).collect();
    actual.sort();
    expected.sort();
    actual == expected
}
